python_modules/dagster/dagster_tests/core_tests/resource_tests/test_struct_config_resources.py

```diff
@@ -430,3 +437,583 @@ def hello_world_asset(writer: WriterResource):
+    match = re.findall(r'Type of "(?:[^"]+)" is "([^"]+)"', stdout)
+    assert match
+    return match
+
+
+def get_mypy_type_output(filename) -> List[str]:
+    stdout = subprocess.check_output(["mypy", filename]).decode("utf-8")
+    match = re.findall(r'note: Revealed type is "([^"]+)"', stdout)
+    assert match
+    return match
+
+
+def test_type_signatures_constructor_nested_resource():
+    with tempfile.TemporaryDirectory() as tempdir:
+        filename = os.path.join(tempdir, "test.py")
+
+        with open(filename, "w") as f:
+            f.write(
+                """
+from dagster._config.structured_config import Resource
+
+class InnerResource(Resource):
+    a_string: str
+
+class OuterResource(Resource):
+    inner: InnerResource
+    a_bool: bool
+
+reveal_type(InnerResource.__init__)
+reveal_type(OuterResource.__init__)
+
+my_outer = OuterResource(inner=InnerResource(a_string="foo"), a_bool=True)
+reveal_type(my_outer.inner)
+"""
+            )
+
+        pyright_out = get_pyright_reveal_type_output(filename)
+        mypy_out = get_mypy_type_output(filename)
+
+        # Ensure constructor signature is correct (mypy doesn't yet support Pydantic model constructor type hints)
+        assert pyright_out[0] == "(self: InnerResource, a_string: str) -> None"
+        assert (
+            pyright_out[1]
+            == "(self: OuterResource, inner: InnerResource | PartialResource[InnerResource],"
+            " a_bool: bool) -> None"
+        )
+
+        # Ensure that the retrieved type is the same as the type of the resource (no partial)
+        assert pyright_out[2] == "InnerResource"
+        assert mypy_out[2] == "test.InnerResource"
+
+
+def test_type_signatures_config_at_launch():
+    with tempfile.TemporaryDirectory() as tempdir:
+        filename = os.path.join(tempdir, "test.py")
+
+        with open(filename, "w") as f:
+            f.write(
+                """
+from dagster._config.structured_config import Resource
+
+class MyResource(Resource):
+    a_string: str
+
+reveal_type(MyResource.configure_at_launch())
+"""
+            )
+
+        pyright_out = get_pyright_reveal_type_output(filename)
+        mypy_out = get_mypy_type_output(filename)
+
+        # Ensure partial resource is correctly parameterized
+        assert pyright_out[0] == "PartialResource[MyResource]"
+        assert mypy_out[0].endswith("PartialResource[test.MyResource]")
+
+
+def test_type_signatures_constructor_resource_dependency():
+    with tempfile.TemporaryDirectory() as tempdir:
+        filename = os.path.join(tempdir, "test.py")
+
+        with open(filename, "w") as f:
+            f.write(
+                """
+from dagster._config.structured_config import Resource, ResourceDependency
+
+class StringDependentResource(Resource):
+    a_string: ResourceDependency[str]
+
+reveal_type(StringDependentResource.__init__)
+
+my_str_resource = StringDependentResource(a_string="foo")
```

Comment:
These tests are supppper slow. I think we should disable them by default locally and additionally, run them in their own test suite to keep the `core_test` time reasonable.

Reply:
Gated behind `-m 'typesignature'` mark. They'll run separately in CI now & only run locally with that flag.